doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,223 @@ entirely in Julia. You can read their source and see precisely what they
+expression depending on the types passed to it:
+
+.. doctest::
+
+   julia> @generated function bar(x)
+              if x <: Integer
+                  return :(x^2)
+              else
+                  return :(x)
+              end
+          end
+    bar (generic function with 1 method)
+
+    julia> bar(4)
+    16
+    julia> bar("baz")
+    "baz"
+
+(although of course this contrived example is easily implemented using
+multiple dispatch...)
+
+We can, of course, abuse this to produce some interesting behavior::
+
+   julia> @generated function baz(x)
+              if rand() < .9
+                  return :(x^2)
+              else
+                  return :("boo!")
+              end
+          end
+
+Since the body of the generated function is non-deterministic, its behavior
+is undefined; the expression returned on the *first* invocation will be
+used for *all* subsequent invocations with the same type (again, with the
+exception covered by the disclaimer above). When we call the generated
+function with ``x`` of a new type, ``rand()`` will be called again to
+see which method body to use for the new type. In this case, for one
+*type* out of ten, ``baz(x)`` will return the string ``"boo!"``.
+
+*Don't copy these examples!*
+
+These examples are hopefully helpful to illustrate how generated functions
+work, both in the definition end and at the call site; however, *don't
+copy them*, for the following reasons:
+
+* the `foo` function has side-effects, and it is undefined exactly when,
+  how often or how many times these side-effects will occur
+* the `bar` function solves a problem that is better solved with multiple
+  dispatch - defining `bar(x) = x` and `bar(x::Integer) = x^2` will do
+  the same thing, but it is both simpler and faster.
+* the `baz` function is pathologically insane
+
+Instead, now that we have a better understanding for how generated functions
+work, let's use them to build some more advanced functionality...
+
+An advanced example
+~~~~~~~~~~~~~~~~~~~
+
+Julia's base library has a function ``sub2ind`` function to calculate a
+linear index into an n-dimensional array, based on a set of n multilinear
+indices - in other words, to calculate the index ``i`` that can be used to
+index into an array ``A`` using ``A[i]``, instead of ``A[x,y,z,...]``. One
+possible implementation is the following::
+
+    function sub2ind_loop(dims::NTuple{N}, I::Integer...)
+        ind = I[N] - 1
+        for i = N-1:-1:1
+            ind = I[i]-1 + dims[i]*ind
+        end
+        return ind + 1
+    end
+
+The same thing can be done using recursion::
+
+    sub2ind_rec(dims::()) = 1
+    sub2ind_rec(dims::(),i1::Integer, I::Integer...) =
+        i1==1 ? sub2ind_rec(dims,I...) : throw(BoundsError())
+    sub2ind_rec(dims::(Integer,Integer...), i1::Integer) = i1
+    sub2ind_rec(dims::(Integer,Integer...), i1::Integer, I::Integer...) =
+        i1 + dims[1]*(sub2ind_rec(tail(dims),I...)-1)
```

Comment:
These are now out of date due to #10380.  Needs to be `Tuple{}` and `Tuple{Integer, Vararg{Integer}}`, etc.